_blog/2024-01-24-using-cronjobs-rust.mdx

```diff
@@ -0,0 +1,294 @@
+
+Then we need to implement a few new things:
+
+- A struct that implements `apalis::prelude::Job`
+- A struct that can be added as a shared data extension to our cron service with a function that will do the actual work we want
+- A function that gets called by `apalis` when work is required to be done (according to the cronjob schedule)
+
+Let’s start with implementing our job struct. Note that when using `apalis` jobs in a cronjob context, they must also implement `From<chrono::DateTime<chrono::Utc>>` (essentially this means they can’t hold any other fields):
+
+```rust
+use chrono::{DateTime, Utc};
+use serde::{Deserialize, Serialize};
+use apalis::prelude::Job;
+
+#[derive(Default, Debug, Clone, Serialize, Deserialize)]
+struct Reminder(DateTime<Utc>);
+
+impl From<DateTime<Utc>> for Reminder {
+   fn from(t: DateTime<Utc>) -> Self {
+       Reminder(t)
+   }
+}
+
+// set up an identifier for apalis
+impl Job for Reminder {
+    const NAME: &'static str = "reminder::DailyReminder";
+}
+```
+
+Now we want to implement the function that will do the actual work when called by `apalis`. We can set this up so that it just says “Hello world from `say_hello_world()`!”:
+
+```rust
+use apalis::prelude::JobContext;
+
+async fn say_hello_world(job: Reminder, ctx: JobContext) {
+    println!("Hello world from `say_hello_world()`!");
+}
+```
+
+However, if you want to add extra variables (for example, a database connection pool) you need to add an extension layer in your service to be able to access them. Let’s have a look at what this struct might look like:
+
+```rust
+#[derive(Clone)]
+struct CronService { 
+    message: String 
+}
+
+impl CronService {
+    fn execute(&self, item: Reminder) {
+        println!("{} from CronService::execute()!", &self.message);
+    }
+}
+```
+
+You can then feed this struct back into the `say_hello_world` function, which uses the `execute()` function:
+
+```rust
+async fn say_hello_world(job: Reminder, ctx: JobContext) {
+    println!("Hello world from send_reminder()!");
+    // this lets you use variables stored in the CronService struct
+    let svc = ctx.data_opt::<CronService>().unwrap();
+    // this executes CronService::execute()
+    svc.execute(job);
+}
+```
+
+### Hooking it all up
+
+Now we’re going to hook the structs we made previously into our main app!
+
+To start with, we need to set up the `PostgresStorage` type so that we can use Postgres for durable job queues. Without durable job queues, our jobs would disappear if our web service has any outages! We can convert this directly from the `PgPool` type stored in our main struct and run the migrations for it like so:
+
+```rust
+#[shuttle_runtime::async_trait]
+impl shuttle_runtime::Service for MyService {
```

Comment:
I think I would explain a bit more what `Service` is for Shuttle and link some documentation.